### DISCUSSION TITLE
IDE compatibility with `getData()`

### DISCUSSION THREAD

**Discussion by HandsomeStrife** [spatie/laravel-data#57] [Q&A]
Hey - am loving the Data package, and currently using it in conjunction with the event sourcing package.

However, the issue I'm facing at the minute stems from using the `getData` function off of Models, via the `WithData` trait; namely as I'm type hinting my methods to expect the data class I'd expect, but with PHPStorm I'm seeing warnings such as:

> Expected parameter of type '\Domain\Team\Data\TeamData', '\Spatie\LaravelData\Data' provided

due to the fact that the getData class specifies Data being returned. Probably a needless example, but just to show a code snippet/explanation of what's going on:

```php
class TeamData extends Data { ... }

class Team extends Model
{
    use WithData;
    public $dataClass = TeamData::class;
}

class TeamRepository
{
    public function doThingWithTeam(TeamData $data) { ... }
}

$team = Team::find(1);
(new TeamRepository)->doThingWithTeam($team->getData()); // shows the warning
```

Currently I'm getting around these warnings in the IDE by defining the data in a docblock, or by wrapping it in ::from e.g.

```php
/** @var TeamData $teamData */
$teamData = $team->getData(); 
// or
$teamData = TeamData::from($team);
```

I wondered if anyone had come across a similar issue and knew of a better solution? I'd much rather use `->getData()` than `::from` as I feel it's a little nicer, but if it's the tradeoff then that's fine :)

Thanks!

**Comment by rubenvanassche** [comment]
Hi @danives,

Hmm, I don't think such a thing is possible at the moment, sadly enough 😢
